editor/debugger/editor_expression_evaluator.cpp

```diff
@@ -0,0 +1,142 @@
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "editor_expression_evaluator.h"
+
+#include "core/debugger/debugger_marshalls.h"
+#include "core/io/marshalls.h"
+#include "core/object/callable_method_pointer.h"
+#include "core/os/os.h"
+#include "editor/debugger/script_editor_debugger.h"
+#include "editor/editor_log.h"
+#include "editor/editor_node.h"
+#include "editor/editor_scale.h"
+#include "editor/editor_settings.h"
+
+void EditorExpressionEvaluator::evaluate() {
+	String expression = expression_input->get_text();
+
+	if (expression.is_empty()) {
+		return;
+	}
+
+	if (editor_debugger->is_session_active()) {
+		EditorNode::get_log()->add_message("Evaluate Expression: " + expression, EditorLog::MSG_TYPE_EDITOR);
+
+		Array expr_data;
+		expr_data.push_back(expression);
+		expr_data.push_back(editor_debugger->get_stack_script_frame());
+
+		Array expr_msg;
+		expr_msg.push_back("evaluate");
+		expr_msg.push_back(Thread::get_caller_id());
+		expr_msg.push_back(expr_data);
+		editor_debugger->get_peer()->put_message(expr_msg);
+
+		expression_input->clear();
+	}
+}
+
+void EditorExpressionEvaluator::clear() {
+	inspector->clear_stack_variables();
+}
+
+void EditorExpressionEvaluator::set_editor_debugger(ScriptEditorDebugger *p_editor_debugger) {
+	editor_debugger = p_editor_debugger;
+}
+
+void EditorExpressionEvaluator::add_value(const Array &p_array) {
+	inspector->add_stack_variable(p_array, 0);
+	inspector->set_v_scroll(0);
+	inspector->set_h_scroll(0);
+}
+
+void EditorExpressionEvaluator::_remote_object_selected(ObjectID p_id) {
+	editor_debugger->emit_signal(SNAME("remote_object_requested"), p_id);
+}
+
+void EditorExpressionEvaluator::_on_expression_input_submitted(String p_expression) {
+	evaluate();
+}
+
+void EditorExpressionEvaluator::_on_project_run() {
+	if (clear_on_run_checkbox->is_pressed()) {
+		inspector->clear_stack_variables();
+	}
+}
+
+void EditorExpressionEvaluator::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_READY: {
+			EditorNode::get_singleton()->connect("project_run", callable_mp(this, &EditorExpressionEvaluator::_on_project_run));
```

Comment:
Can't this be connected in the constructor? I think EditorNode already exists then.

Reply:
Idk... for your latest requests, I added more signal handlers too on the ready notification to make sure.